File: api/src/services/users.service.ts
/**
 * User directory service. Seeds initial App Owners from config on first run,
 * derives the human-facing authentication method, and enforces that only
 * App Owners maintain users (checked again at the route layer).
 */
import { v4 as uuid } from 'uuid';
import { getRepository } from '../persistence/factory';
import { config } from '../config';
import { AuthProvider, User, UserRole, ALL_ROLES } from '../models';

const repo = () => getRepository<User>('users');

function methodFor(provider: AuthProvider): User['authenticationMethod'] {
  return provider === 'email-otp' ? 'OTP' : 'Entra ID';
}

function normalizeEmail(email: string): string {
  return email.trim().toLowerCase();
}

export async function ensureSeedUsers(): Promise<void> {
  for (const seed of config.seedUsers) {
    const email = normalizeEmail(seed.email);
    const existing = await getByEmail(email);
    if (existing) continue;
    const now = new Date().toISOString();
    await repo().upsert({
      id: uuid(),
      email,
      name: seed.name,
      role: seed.role as UserRole,
      provider: seed.provider as AuthProvider,
      authenticationMethod: methodFor(seed.provider as AuthProvider),
      createdAt: now,
      updatedAt: now,
    });
  }
}

export async function getByEmail(email: string): Promise<User | undefined> {
  const target = normalizeEmail(email);
  const rows = await repo().find((u) => u.email === target);
  return rows[0];
}

export async function getById(id: string): Promise<User | undefined> {
  return repo().getById(id);
}

export async function list(): Promise<User[]> {
  return (await repo().getAll()).sort((a, b) => a.email.localeCompare(b.email));
}

export interface UpsertUserInput {
  email: string;
  name: string;
  role: UserRole;
  provider: AuthProvider;
  disabled?: boolean;
}

export async function createOrUpdate(input: UpsertUserInput): Promise<User> {
  if (!ALL_ROLES.includes(input.role)) {
    throw Object.assign(new Error(`Invalid role: ${input.role}`), { status: 400 });
  }
  const email = normalizeEmail(input.email);
  const existing = await getByEmail(email);
  const now = new Date().toISOString();
  const user: User = {
    id: existing?.id ?? uuid(),
    email,
    name: input.name,
    role: input.role,
    provider: input.provider,
    authenticationMethod: methodFor(input.provider),
    disabled: input.disabled ?? existing?.disabled ?? false,
    createdAt: existing?.createdAt ?? now,
    updatedAt: now,
  };
  return repo().upsert(user);
}

export async function remove(id: string): Promise<boolean> {
  return repo().delete(id);
}

/** Resolve a user for an authenticated email, defaulting new internal users. */
export async function resolveForLogin(
  email: string,
  name: string,
  provider: AuthProvider,
): Promise<User> {
  const existing = await getByEmail(email);
  if (existing) return existing;
  // Unknown users default to business_user; App Owner can elevate later.
  return createOrUpdate({ email, name, role: 'business_user', provider });
}

File: api/app/users_service.py
"""User directory service (mirror of users.service.ts)."""
from typing import Dict, List, Optional

from .config import CONFIG
from .persistence import get_repository
from .rbac import ALL_ROLES
from .util import new_id, now_iso


def _repo():
    return get_repository("users")


def _method_for(provider: str) -> str:
    return "OTP" if provider == "email-otp" else "Entra ID"


def _normalize_email(email: str) -> str:
    return email.strip().lower()


def ensure_seed_users() -> None:
    for seed in CONFIG["seedUsers"]:
        email = _normalize_email(seed["email"])
        if get_by_email(email):
            continue
        now = now_iso()
        _repo().upsert(
            {
                "id": new_id(),
                "email": email,
                "name": seed["name"],
                "role": seed["role"],
                "provider": seed["provider"],
                "authenticationMethod": _method_for(seed["provider"]),
                "createdAt": now,
                "updatedAt": now,
            }
        )


def get_by_email(email: str) -> Optional[Dict]:
    target = _normalize_email(email)
    rows = _repo().find(lambda u: u.get("email") == target)
    return rows[0] if rows else None


def get_by_id(user_id: str) -> Optional[Dict]:
    return _repo().get_by_id(user_id)


def list_users() -> List[Dict]:
    return sorted(_repo().get_all(), key=lambda u: u.get("email", ""))


def create_or_update(
    *, email: str, name: str, role: str, provider: str, disabled: Optional[bool] = None
) -> Dict:
    if role not in ALL_ROLES:
        from .errors import ApiError

        raise ApiError(400, f"Invalid role: {role}")
    norm = _normalize_email(email)
    existing = get_by_email(norm)
    now = now_iso()
    user = {
        "id": existing["id"] if existing else new_id(),
        "email": norm,
        "name": name,
        "role": role,
        "provider": provider,
        "authenticationMethod": _method_for(provider),
        "disabled": disabled if disabled is not None else (existing.get("disabled") if existing else False),
        "createdAt": existing["createdAt"] if existing else now,
        "updatedAt": now,
    }
    return _repo().upsert(user)


def remove(user_id: str) -> bool:
    return _repo().delete(user_id)


def resolve_for_login(email: str, name: str, provider: str) -> Dict:
    existing = get_by_email(email)
    if existing:
        return existing
    return create_or_update(email=email, name=name, role="business_user", provider=provider)
